handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -96,10 +112,40 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
             }
             this.hostname = hostname;
 
-            // the mapping will return default context when this.hostname is null
-            selectedContext = mapping.map(hostname);
+            Future<SslContext> future = mapping.map(hostname, ctx.executor().<SslContext>newPromise());
+            if (future.isDone()) {
+                if (future.isSuccess()) {
+                    selectContext(ctx, future);
+                } else {
+                    throw new DecoderException("failed to get the SslContext for " + hostname, future.cause());
+                }
+            } else {
+                suppressRead = true;
```

Comment:
now that we have state, do we need to prevent this state from being "overwritten" if a lookup operation is in progress?

For example what if decode is called multiple times, and is this a realistic scenario?